hail/src/main/scala/is/hail/types/encoded/EArray.scala

```diff
@@ -100,22 +100,66 @@ final case class EArray(val elementType: EType, override val required: Boolean =
 
     val len = cb.newLocal[Int]("len", in.readInt())
     val array = cb.newLocal[Long]("array", arrayType.allocate(region, len))
+    val elemAddr = cb.newLocal[Long]("elemAddr", arrayType.firstElementOffset(array, len))
     arrayType.storeLength(cb, array, len)
 
     val i = cb.newLocal[Int]("i")
+    val mbyteOffset = cb.newLocal[Long]("mbyteOffset", array + arrayType.lengthHeaderBytes)
+    val mbyte = cb.newLocal[Byte]("mbyte", 0.toByte)
+    val presentBitsLong = cb.newLocal[Long]("presentBitsLong", 0)
+    val inBlockIndexToPresentValue = cb.newLocal[Int]("inBlockIndexToPresentValue", 0)
     val readElemF = elementType.buildInplaceDecoder(arrayType.elementType, cb.emb.ecb)
 
     if (!elementType.required)
       cb += in.readBytes(region, array + const(arrayType.lengthHeaderBytes), arrayType.nMissingBytes(len))
 
-    cb.forLoop(cb.assign(i, 0), i < len, cb.assign(i, i + 1), {
-      val elemAddr = cb.memoize(arrayType.elementOffset(array, len, i))
-      if (elementType.required)
-        readElemF(cb, region, elemAddr, in)
-      else
-        cb.ifx(arrayType.isElementDefined(array, i),
-          readElemF(cb, region, elemAddr, in))
-    })
+    val LONG_MASK_ALL_BUT_FIRST = const(0x7fffffffffffffffL)
+
+    cb.assign(i, 0)
+    if (elementType.required) {
+      cb.forLoop({}, i + 64 < len, cb.assign(i, i + 64), {
+        for (_ <- 0 to 63) {
+          readElemF(cb, region, elemAddr, in)
+          cb.assign(elemAddr, arrayType.nextElementAddress(elemAddr))
+        }
+      })
+    } else {
+      cb.forLoop({}, i + 64 < len, cb.assign(i, i + 64), {
+        cb.assign(presentBitsLong, ~Region.loadLong(mbyteOffset))
+        cb.assign(mbyteOffset, mbyteOffset + 8)
+        cb.assign(inBlockIndexToPresentValue, presentBitsLong.numberOfLeadingZeros)
+
+        cb.whileLoop(inBlockIndexToPresentValue < 64, {
+          cb.assign(elemAddr,
+            arrayType.elementOffset(array, len, i + inBlockIndexToPresentValue))
+          readElemF(cb, region, elemAddr, in)
+
+          cb.assign(inBlockIndexToPresentValue,
+            (presentBitsLong & (LONG_MASK_ALL_BUT_FIRST >>> inBlockIndexToPresentValue)).numberOfLeadingZeros)
+        })
+      })
+      cb.assign(elemAddr, arrayType.elementOffset(array, len, i))
```

Comment:
Ah, I think you have the bit ordering wrong. We arrange bytes left-to-right, and bits within a byte right-to-left (which I think is right; the index 0 bit should be the least-significant bit). It might be worth changing PCanonicalArray to group missing bits by 64 rather than 8, to make tricks like this faster. I don't expect reading a byte is any faster than reading a long. But in the meantime, we can just use `Long.reverseBytes`.

Note that the right-to-left bit order allows a faster way to do your original approach:

```scala
    def reverseBytes(l: Code[Long]): Code[Long] =
      Code.invokeStatic1[java.lang.Long, Long, Long]("reverseBytes", l)
```
```suggestion
      cb.assign(blockAddr, arrayType.firstElementOffset(array, len))
      val endPtr = cb.memoize(arrayType.elementOffset(array, len, len - 64))
      // need to make nextElementAddress(addr, offset) or something
      cb.forLoop({}, elemAddr < endPtr, cb.assign(blockAddr, arrayType.nextElementAddress(array, offset=64)), {
        cb.assign(presentBitsLong, ~reverseBytes(Region.loadLong(mbyteOffset)))
        cb.assign(mbyteOffset, mbyteOffset + 8)

        cb.whileLoop(presentBitsLong != 0, {
          cb.assign(inBlockIndexToPresentValue, presentBitsLong.numberOfTrailingZeros)
          val cb.assign(elemAddr,
            arrayType.nextElementOffset(blockAddr, inBlockIndexToPresentValue))
          readElemF(cb, region, elemAddr, in)

          // unset rightmost set bit
          cb.assign(presentBitsLong, presentBitsLong & (presentBitsLong - 1))
        })
      })
```

Reply:
killer find, this would've taken me quite a while to sort out.